### PR TITLE
Core suite 113.33 + 4.03.0 support

### PR DESCRIPTION
**Pull request by trefis**
This corresponds to the same version of our libraries as for PR #5928 (fixes included) but updated to compile and work with OCaml 4.03.

We usually send our PR with urls pointing to our git repos and update to tarballs on ocaml.janestreet.com when everything looks ok. We did not do this here since the git repos require oasis (we do not commit the generated files) which is not yet available on opam for 4.03; so we directly offer to test against the tarball.

Note: we do not plan to do any more work on this for new platforms (alpine or other) than what was done for PR #5928 but will try and fix all 4.03-related problems.

Thanks

### CODE REVIEW

**Comment by avsm** [comment]
Bulk build against 4.02.3 and 4.03.0_trunk triggered; will be in https://github.com/avsm/opam-pr-build-results/blob/6094/README.md in about an hour.  the current results there are out-of-disk space due to infrastructure error.

**Comment by trefis** [comment]
```
... opam switch 4.03.0_trunk ...
[ERROR] "4.03.0_trunk" is not a valid compiler.
```

:(

**Comment by avsm** [comment]
oops, typo fail. fixing it...

**Comment by avsm** [comment]
ok i did a smoke test with sexplib and there's a good clean build on 4.03 4.02 and flambda. Now starting core_extended jenga and async_ssl...

https://github.com/avsm/opam-pr-build-results/blob/6094/README.md

**Comment by avsm** [comment]
Lots of 

```
- + /home/opam/.opam/4.03.0+trunk/bin/ocamlfind ocamldep -package variantslib -package typerep -package sexplib -package result -package ppx_inline_test.runtime-lib -package ppx_expect.collector -package ppx_bench.runtime-lib -package ppx_assert.runtime-lib -package num -package fieldslib -package bin_prot -pp 'ppx-jane -dump-ast -inline-test-drop -bench-drop -inline-test-lib fake_for_drop' -predicates ppx_driver -modules src/core_map.mli > src/core_map.mli.depends
- File "_none_", line 1:
- attribute not expected here, Ppx_core.Std.Attribute needs updating!
- Command exited with code 2.
```

**Comment by trefis** [comment]
I haven't seen any of these, all the builds I saw failing were due to using `ppx_expect.113.33.01` instead of `ppx_expect.113.33.00+4.03`.
I'll let you know when I update the PR with fixed constraints.

**Comment by avsm** [comment]
@trefis i restarted the build last night with the very latest 4.03+trunk container, so I guess that there was a ppx_tools update that fixed it

**Comment by trefis** [comment]
Hum, might be...
Anyway I just rebased the PR and added a commit fixing the ocaml version constraints for the packages working only on 4.02.

**Comment by avsm** [comment]
build restarted

**Comment by avsm** [comment]
@trefis we're back to the
```
- + /home/opam/.opam/4.03.0+trunk/bin/ocamlfind ocamldep -package variantslib -package typerep -package sexplib -package result -package ppx_inline_test.runtime-lib -package ppx_expect.collector -package ppx_bench.runtime-lib -package ppx_assert.runtime-lib -package num -package fieldslib -package bin_prot -pp 'ppx-jane -dump-ast -inline-test-drop -bench-drop -inline-test-lib fake_for_drop' -predicates ppx_driver -modules src/core_map.mli > src/core_map.mli.depends
- File "_none_", line 1:
- attribute not expected here, Ppx_core.Std.Attribute needs updating!
- Command exited with code 2.
- + ocamlfind ocamlopt -package unix -package ocamlbuild -linkpkg -package ppx_driver.ocamlbuild myocamlbuild.ml /home/opam/.opam/4.03.0+trunk/lib/ocamlbuild/ocamlbuild.cmx -o myocamlbuild
- findlib: [WARNING] Interface topdirs.cmi occurs in several directories: /home/opam/.opam/4.03.0+trunk/lib/ocaml/compiler-libs, /home/opam/.opam/4.03.0+trunk/lib/ocaml
```

errors

**Comment by trefis** [comment]
Ah! I was testing with 4.03.0+beta1 internally, I will upgrade to 4.03.0+trunk.

**Comment by trefis** [comment]
Alright, ppx_core should be fixed now!

**Comment by avsm** [comment]
Build finished: https://github.com/avsm/opam-pr-build-results/blob/6094/README.md

**Comment by trefis** [comment]
It looks good to me (the failures seem somewhat random: missing g++, installation of distribution package failure, etc.).
Do you agree?

**Comment by avsm** [comment]
Ah the g++ depext is because I merged in changes whihc arent in the latest.  Any chance you could sync -- e.g. in re2

```
 depexts: [
 [ ["fedora"] ["gcc-c++"] ]
 [ ["oraclelinux"] ["gcc-c++"] ]
]

**Comment by avsm** [comment]
testing on freebsd now

**Comment by trefis** [comment]
> Any chance you could sync -- e.g. in re2 [...]

I guess we want that in both "versions" of 113.33.
Should I updated earlier versions of the package as well?

**Comment by avsm** [comment]
@trefis - yeah, might be worth creating a `conf-g++` package with the depexts so that there can just be one dependency

**Comment by avsm** [comment]
freebsd core still has

```
#=== ERROR while installing core.113.33.00 ====================================#
# opam-version 1.2.2
# os           freebsd
# command      gmake
# path         /root/.opam/system/build/core.113.33.00
# compiler     system (4.02.3)
# exit-code    2
# env-file     /root/.opam/system/build/core.113.33.00/core-3796-8a4173.env
# stdout-file  /root/.opam/system/build/core.113.33.00/core-3796-8a4173.out
# stderr-file  /root/.opam/system/build/core.113.33.00/core-3796-8a4173.err
### stdout ###
# [...]
# /root/.opam/system/bin/ocamlfind ocamlc -g -I src -package core_kernel -c src/bigstring_stubs.c
# + /root/.opam/system/bin/ocamlfind ocamlc -g -I src -package core_kernel -c src/bigstring_stubs.c
# src/bigstring_stubs.c:34:10: fatal error: 'endian.h' file not found
# #include <endian.h>
#          ^
# 1 error generated.
# Command exited with code 2.
```

**Comment by avsm** [comment]
@trefis ping on the FreeBSD issue; I think it would be very good to fix that one if possible as it was a regression.

**Comment by trefis** [comment]
Yes, I started working on this earlier then forgot.
I'll have something for you soon.

**Comment by trefis** [comment]
Done!

I'll make another PR to fix core with 4.02 soon.

File: packages/conf-g++/conf-g++.0.1/opam [hunk 1]
```diff
@@ -0,0 +1,17 @@
+opam-version: "1.2"
+maintainer: "trefis@janestreet.com"
+homepage: "http://gcc.gnu.org"
+authors: "http://www.gnuplot.info/faq/faq.html"
+bug-reports: "https://github.com/ocaml/opam-repository/issues"
+dev-repo: "https://github.com/ocaml/opam-repository.git"
+license: "https://gcc.gnu.org/install/gfdl.html"
+build: [
+  ["which" "g++"]
```
Comment (avsm) [inline packages/conf-g++/conf-g++.0.1/opam]: this actually needs to run `which c++` since the underlying compiler may be clang (on freebsd and osx)

Comment (avsm) [inline packages/conf-g++/conf-g++.0.1/opam]: so  perhaps package should be `conf-c++` to  be more portable

**Comment by avsm** [comment]
Sigh, 4.03 is still a moving target...

```
# findlib: [WARNING] Interface topdirs.cmi occurs in several directories: /root/.opam/4.03.0+trunk/lib/ocaml, /root/.opam/4.03.0+trunk/lib/ocaml/compiler-libs
# File "src/ppx_deriving_main.ml", line 19, characters 37-49:
# Error: This variant pattern is expected to have type Parsetree.constant
#        The constructor Const_string does not belong to type Parsetree.constant
# Hint: Did you mean Pconst_string?
```

with a fresh 4.03.0+trunk

**Comment by avsm** [comment]
(that was ppx_deriving)

**Comment by trefis** [comment]
Hum, ping @whitequark then?

Btw, I initially prepared this release against 4.03+beta1, with the understanding that people were using this to test 4.03.
4.03+trunk is nice but indeed a moving target and I'm not sure who is using it? Is another beta planned sometime soon?

File: packages/conf-g++/conf-g++.0.1/opam (hunk 1 above)
Comment (trefis) [inline packages/conf-g++/conf-g++.0.1/opam]: I reverted that and just added the depexts to re2.
I'll let someone who knows this stuff do the conf- package properly.

**Comment by avsm** [comment]
> Btw, I initially prepared this release against 4.03+beta1, with the understanding that people were using this to test 4.03.

It seems to have diverged so much that there's no real point testing against beta1 at this point.  @damiendoligez, would you be planning a beta2 cut soon so that we can get a version of Core merged against it?

**Comment by whitequark** [comment]
ppx_deriving was recently updated for 4.03; I bumped the major along the way. Might it be that Core is picking up an old version?

**Comment by trefis** [comment]
We depend on `ppx_deriving >= 3.0` so I think we should be fine.
I guess your version works with 4.03.0+beta1 and not with trunk which is what Anil is testing against here?

**Comment by whitequark** [comment]
I've definitely implemented `Pconst_string` so probably not. I'll take a look tomorrow

**Comment by trefis** [comment]
Indeed, so we most likely picked the wrong version...
@avsm is the full log accessible somewhere?

**Comment by avsm** [comment]
Looks like internal solver badness then; 

```
=-=- Error report -=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
The following actions were aborted
  ∗  install core              113.33.01+4.03
  ∗  install core_extended     113.33.00+4.03
  ∗  install core_kernel       113.33.01+4.03
  ∗  install patdiff           113.33.00     
  ∗  install patience_diff     113.33.00     
  ∗  install ppx_assert        113.33.00     
  ∗  install ppx_bin_prot      113.33.00+4.03
  ∗  install ppx_compare       113.33.00+4.03
  ∗  install ppx_custom_printf 113.33.00+4.03
  ∗  install ppx_enumerate     113.33.00+4.03
  ∗  install ppx_expect        113.33.00+4.03
  ∗  install ppx_fields_conv   113.33.00+4.03
  ∗  install ppx_jane          113.33.00     
  ∗  install ppx_sexp_conv     113.33.00+4.03
  ∗  install ppx_sexp_message  113.33.00+4.03
  ∗  install ppx_sexp_value    113.33.00+4.03
  ∗  install ppx_type_conv     113.33.00+4.03
  ∗  install ppx_typerep_conv  113.33.00+4.03
  ∗  install ppx_variants_conv 113.33.00+4.03
  ∗  install re2               113.33.00+4.03
  ∗  install textutils         113.33.00     
The following actions failed
  ∗  install ppx_deriving 3.2
The following changes have been performed
  ∗  install bin_prot        113.33.00+4.03
  ∗  install conf-g++        0.1           
  ∗  install conf-libpcre    1             
  ∗  install conf-m4         1             
  ∗  install conf-ncurses    1+system      
  ∗  install conf-pkg-config 1.0           
  ∗  install conf-which      1             
  ∗  install fieldslib       113.24.00     
  ∗  install ocamlbuild      0.9.1         
  ∗  install ocamlfind       1.6.2         
  ∗  install pcre            7.1.3         
  ∗  install ppx_bench       113.33.00+4.03
  ∗  install ppx_core        113.33.00+4.03
  ∗  install ppx_driver      113.33.00+4.03
  ∗  install ppx_fail        113.33.00+4.03
  ∗  install ppx_here        113.33.00     
  ∗  install ppx_inline_test 113.33.00+4.03
  ∗  install ppx_let         113.33.00+4.03
  ∗  install ppx_optcomp     113.33.00+4.03
  ∗  install ppx_pipebang    113.33.00+4.03
  ∗  install ppx_tools       4.03.0        
  ∗  install re              1.2.2         
  ∗  install result          1.2           
  ∗  install sexplib         113.33.00+4.03
  ∗  install typerep         113.24.00     
  ∗  install variantslib     113.24.00     

```

ppx_deriving.3.3 installs fine. Trying again with online solver on freebsd...

**Comment by whitequark** [comment]
> I bumped the major along the way.

Sorry, it was late yesterday and I confused myself. I did not yet release ppx_deriving 4.0, but I did release 3.3 with 4.03.0 support. Anyway, trying to install 3.2 will rightly fail. We need to add the `available` metadata to the repository...

**Comment by damiendoligez** [comment]
I'll release a beta2 today or tomorrow.

**Comment by avsm** [comment]
containers updated to reflect slightly after beta2, @trefis

**Comment by trefis** [comment]
I tested with 4.03.0+beta2, things compile on my side.
I also updated the packages to the latest minor releases of 113.33.

**Comment by avsm** [comment]
followon fix for fedora in #6222

**Pull request merged**